examples/M4_competition/FourTheta.py

```diff
@@ -0,0 +1,342 @@
+import math
+from typing import Optional, List, Callable
+
+import numpy as np
+import statsmodels.tsa.holtwinters as hw
+from itertools import product
+
+from darts.utils.statistics import check_seasonality, extract_trend_and_seasonality, remove_from_series
+from darts.utils import _build_tqdm_iterator
+from darts.models.forecasting_model import UnivariateForecastingModel
+from darts.logging import raise_if_not, get_logger
+from darts import TimeSeries, TrendMode, SeasonalityMode, ModelMode
+
+from sklearn.metrics import mean_absolute_error as mae
+from darts.models import AutoARIMA
+
+logger = get_logger(__name__)
+
+
+def backtest_gridsearch(model_class: type,
+                        parameters: dict,
+                        train_series: TimeSeries,
+                        metric: Callable[[np.ndarray, np.ndarray], float] = mae,
+                        verbose=False):
+
+    model = model_class()
+    raise_if_not(hasattr(model, "fitted_values"), "The model must have a fitted_values attribute"
+                                                  " to compare with the train TimeSeries", logger)
+
+    min_error = float('inf')
+    best_param_combination = {}
+
+    # compute all hyperparameter combinations from selection
+    params_cross_product = list(product(*parameters.values()))
+
+    # iterate through all combinations of the provided parameters and choose the best one
+    iterator = _build_tqdm_iterator(params_cross_product, verbose)
+    for param_combination in iterator:
+        param_combination_dict = dict(list(zip(parameters.keys(), param_combination)))
+        model = model_class(**param_combination_dict)
+        model.fit(train_series)
+        # Takes too much time to create a TimeSeries
+        # Overhead: 2-10 ms in average
+        # fitted_values = TimeSeries.from_times_and_values(train_series.time_index(), model.fitted_values)
+        error = metric(model.fitted_values, train_series.univariate_values())
+        if error < min_error:
+            min_error = error
+            best_param_combination = param_combination_dict
+    logger.info('Chosen parameters: ' + str(best_param_combination))
+    return model_class(**best_param_combination)
+
+
+class FourTheta(UnivariateForecastingModel):
```

Comment:
This class is already included in the previous PR and merged, right? Or is it some different implementation?

Reply:
The difference is that it uses a modified version of gridsearch that do not create a TimeSeries.
Thus the execution time of the notebook drops from 1h to 5-10 minutes

Reply:
I know it is a pity that using TimeSeries increases the gridsearch time, but if we release this we cannot have a separate implementation, and we should really use the implementation in `darts/models`, even if it takes much longer to run.